### PR TITLE
enable multitouch

### PR DESCRIPTION
**Pull request by p-an**
Hi,
this enables multitouch support. Could you, please, merge it?

Thank you

Pavel

### CODE REVIEW

File: android/opengl/system/gralloc/gralloc.cpp [hunk 1]
```diff
@@ -688,7 +688,7 @@ static int gralloc_lock(gralloc_module_t const* module,
         if (sw_read) {
             D("gralloc_lock read back color buffer %d %d\n", cb->width, cb->height);
             rcEnc->rcReadColorBuffer(rcEnc, cb->hostHandle,
-                    0, 0, cb->width, cb->height, GL_RGBA, GL_UNSIGNED_BYTE, cpu_addr);
+                    0, 0, cb->width, cb->height, cb->glFormat, cb->glType, cpu_addr);
```
Comment (morphis) [inline android/opengl/system/gralloc/gralloc.cpp]: This doesn't belong in this PR.

Comment (p-an) [reply android/opengl/system/gralloc/gralloc.cpp]: Done. I am sorry about this.

Comment (p-an) [reply android/opengl/system/gralloc/gralloc.cpp]: Sorry again. I did not notice you are asking also about the other PR. I`ll take a look at it.

Comment (p-an) [reply android/opengl/system/gralloc/gralloc.cpp]: Hi,
as you have probably noticed I am not too familiar with github. So please forgive me a dumb question. I understand what you want me to do from the code perspective, but I am quite lost with what I am supposed to do from the github prespective. As I understand it, the simplest way would be to merge #1073, close this PR and create a new one based on the merged code. Or is there some better way?

I have also one concern about the code changes in #1073. I am no expert on the touch protocol, but what is in that PR doesn`t seem to be consistent with what I found (and followed in these PR code changes) here https://www.kernel.org/doc/Documentation/input/multi-touch-protocol.txt

e.g. - please compare:

```
Protocol Example B
------------------
Here is what a minimal event sequence for a two-contact touch would look
like for a type B device:

   ABS_MT_SLOT 0
   ABS_MT_TRACKING_ID 45
   ABS_MT_POSITION_X x[0]
   ABS_MT_POSITION_Y y[0]
   ABS_MT_SLOT 1
   ABS_MT_TRACKING_ID 46
   ABS_MT_POSITION_X x[1]
   ABS_MT_POSITION_Y y[1]
   SYN_REPORT
```
 with:
```
touch_events.push_back({EV_ABS, ABS_MT_TRACKING_ID, static_cast<std::int32_t>(emulated_touch_id_)});
touch_events.push_back({EV_ABS, ABS_MT_SLOT, 0});
touch_events.push_back({EV_KEY, BTN_TOUCH, 1});
touch_events.push_back({EV_KEY, BTN_TOOL_FINGER, 1});
touch_events.push_back({EV_ABS, ABS_MT_POSITION_X, x});
touch_events.push_back({EV_ABS, ABS_MT_POSITION_Y, y});
```
According to the kernel doc, the ABS_MT_SLOT should be sent before ABS_MT_TRACKING_ID. Also sending ABS_MT_TOUCH_MAJOR and ABS_MT_TOUCH_MINOR seems to be redundant, since there is no actual finger with these properties. But I my be wrong and those fake values are useful for something.

I suppose that this doesn`t block me from merging my changes with #1073 but I replaced similar touch protocol implemetation with the multitouch implementation. I suppose that it would be nice to have it consistent.

Pavel

Comment (salvprest) [reply android/opengl/system/gralloc/gralloc.cpp]: I had simply reused the code on the touch part, so if you can rewrite it in a way that is consistent with the multi-touch protocol and that works, go ahead

Comment (p-an) [reply android/opengl/system/gralloc/gralloc.cpp]: I know, I noticed that. What I don`t know is if my understanding of kernel doc is wrong or the original implementation is wrong. 

I can reimplement #1073 with my approach, it should not be difficult. Do you have any app to test the touch emulation?  I haven`t noticed any difference with those I tired.

Comment (salvprest) [reply android/opengl/system/gralloc/gralloc.cpp]: the original implementation was written by an ubuntu touch developer, I think, so was written with that kind of device in mind.
Try [this game](https://play.google.com/store/apps/details?id=com.RockyHong.BeatStomperAndroid) it works only with touch inputs

Comment (p-an) [reply android/opengl/system/gralloc/gralloc.cpp]: OK, this makes it more confusing. Did you take a look at the kernel doc? As I understand it ABS_MT_SLOT should be definitely before ABS_MT_TRACKING_ID. Or am I wrong?

The BTN_TOUCH and BTN_TOOL_FINGER are meant AFAIK for single touch devices, so for multitouch they have to be removed.

ABS_MT_TOUCH_MAJOR and ABS_MT_TOUCH_MINOR may be useful in some situations, even fake ones. I suppose they don`t do any harm, so they can be there. 

That works for me as test, so I`ll try to merge it.

Comment (awaken1ng) [reply android/opengl/system/gralloc/gralloc.cpp]: BTN_TOUCH (and, I assume, BTN_TOOL_FINGER for compatibility reasons) seems to be used for indicating that user is touching the screen, touch MOUSE_DOWN of sorts

Here's a `getevent -l` snip from my phone
```
# first finger down
/dev/input/event1: EV_ABS       ABS_MT_TRACKING_ID   00009648
/dev/input/event1: EV_ABS       ABS_MT_POSITION_X    000002c8
/dev/input/event1: EV_ABS       ABS_MT_POSITION_Y    0000021a
/dev/input/event1: EV_KEY       BTN_TOUCH            DOWN    
/dev/input/event1: EV_SYN       SYN_REPORT           00000000
# second finger down
/dev/input/event1: EV_ABS       ABS_MT_SLOT          00000001
/dev/input/event1: EV_ABS       ABS_MT_TRACKING_ID   00009649
/dev/input/event1: EV_ABS       ABS_MT_POSITION_X    00000247
/dev/input/event1: EV_ABS       ABS_MT_POSITION_Y    00000363
/dev/input/event1: EV_SYN       SYN_REPORT           00000000
# first finger up
/dev/input/event1: EV_ABS       ABS_MT_SLOT          00000000
/dev/input/event1: EV_ABS       ABS_MT_TRACKING_ID   ffffffff
/dev/input/event1: EV_SYN       SYN_REPORT           00000000
# second finger up
/dev/input/event1: EV_ABS       ABS_MT_SLOT          00000001
/dev/input/event1: EV_ABS       ABS_MT_TRACKING_ID   ffffffff
/dev/input/event1: EV_KEY       BTN_TOUCH            UP      
/dev/input/event1: EV_SYN       SYN_REPORT           00000000
```

Comment (salvprest) [reply android/opengl/system/gralloc/gralloc.cpp]: @p-an  yes from the documetation and by a quick look to the linux souce code ABS_MT_SLOT should be definitely before ABS_MT_TRACKING_ID, but linux can also handle some corner cases for driver that have a more lax implementation. But yes the best thing is to follow the documentation. @awaken1ng BTN_TOUCH, taking another quick look to the source code, seem to be used to "Perform legacy pointer emulation" I don't know if it is necessary, probably not

Comment (p-an) [reply android/opengl/system/gralloc/gralloc.cpp]: Hi,
I was incorrect about the BTN_TOUCH - it is problematic as a flag. I had to remove 
```
touch_->set_key_bit(BTN_TOUCH);
touch_->set_key_bit(BTN_TOOL_FINGER);
``` 
to make multitouch work. Some hints are here https://source.android.com/devices/input/touch-devices

I am not sure about BTN_TOUCH as an event. It may have some purpose, but so far it is working without it.

Here https://github.com/p-an/anbox/commits/pr1073-merge is how I see it - #1073 reimplemented together with multitouch. It may require some cosmetic changes, but it works, with one exception - combining the emulated touches with real ones sometimes fails. But it is probably a problem of the xorg - when the mouse button is down, touches are sent as mouse events.

Comment (salvprest) [reply android/opengl/system/gralloc/gralloc.cpp]: @p-an now that #1073 is merged in master you can reset the head of the repository of this pr, merge with master, and apply the changes from pr1073-merge
